### DISCUSSION TITLE
Data inconsistency between nodes in a cluster

### DISCUSSION THREAD

**Discussion by krivulcik** [ravendb/ravendb#14472] [Q&A] [answered]
After a failure of one of our nodes (node B) in one of our clusters, we have a data inconsistency between our nodes. There are two databases where the documents are not complete on two of the nodes.

The databases have identical Database Change Vector on all nodes.

You can find the output of /databases/databasename/stats endpoint to check the full Database Change Vectors here (the files are formatted and renamed to .txt to allow uploading them to GitHub):
[stats-main-A.txt](https://github.com/ravendb/ravendb/files/9022650/stats-main-A.txt)
[stats-main-B.txt](https://github.com/ravendb/ravendb/files/9022651/stats-main-B.txt)
[stats-main-C.txt](https://github.com/ravendb/ravendb/files/9022652/stats-main-C.txt)

The document count is the following:
A: 3716 (223 documents missing)
B: 3706 (233 documents missing (10 compared to node A) - this is the failed and recovered node)
C: 3939 (all documents)

The missing documents are from the past two months - the earliest I see is from 2022-04-29.

This is the cluster where we investigated the behavior earlier today, and there are suspected network issues in Azure. However, having identical change vectors and inconsistent data is very alarming even in this environment.

Is there any known scenario where such a situation could happen? Is it possible to investigate more closely?

Fortunately, the missing documents are not critical for our operation. Thanks to this, we can afford to have these documents missing for several days if you would be willing and interested in investigating our environment directly. Additionally, I have created exports of the databases in question. If these would be useful and sufficient, I can provide them for your investigation.

**Comment by karmeli87** [comment, marked as answer]
Hi,

What raven version are you running? is that the same version since 2022-04-29?   

Do you have any revisions? the provided stats seems to omit `CountOfRevisionDocuments`.

Are there any Map-Reduce indexes that output the reduce result into a collection? (I recall we had something like this when modifying artificial document directly)

**Reply by krivulcik** [reply]
We were running 5.3.102 from 2022-04-25 - shortly before the earliest missing document. We have updated from an earlier version, either 5.3.100 or 5.3.2, not sure which one.

Then, we updated it to 5.3.103 on 2022-06-08 (the next day after release).

This database doesn't contain any revisions. In the Studio, the stats screen shows: Revision Count: 0.

There are no indexes that would create artificial documents. There are only two simple map indexes.

I didn't realize that we had upgraded only several days before the earliest known instance of a missing document, so this might mean that there was a new issue introduced in 5.3.102 that can cause this behavior.

**Reply by karmeli87** [reply]
In the change vector there are several elements with `A` tag (and few `B`s).
Those extra elements are usually created when
 
- Remove and re-add the node to the database group
- Import revisions/counters
- External replication to this database 

Does any of the above rings a bell?

I also see you use cluster-wide tx, might it happened that those missing docs was created/deleted by the cluster-wide tx?

**Reply by krivulcik** [reply]
Over the years of the cluster operation, we have lost and re-added some nodes, yes. After each loss and recovery, the data was fully recovered, as far as we could see (we didn't notice anything unusual).

Yes, we do use cluster-wide transactions, and:

 - There are only two collections in this database.
 - Only documents in one collection are missing.
 - This collection started using cluster-wide transactions to create documents on 2022-04-13.

This also precedes the earliest missing document by around two weeks.

However:

 - As I mentioned in the first post, we have two databases with this problem.
 - I was working with only one database so far in this discussion because the other contains sensitive data (user account data), didn't have exactly the same change vectors, and the documents get deleted by an automatic process in the background, so it's not easy to work with the database over longer periods of time.
 - In the other database, the missing documents are *not* created in a cluster-wide transaction.

**Reply by karmeli87** [reply]
If you are able to share the database (with the invalid state) for investigation, please send them to support@ravendb.net 
Ideally we would like to have the raw data of the database from all 3 nodes (you can exclude the indexes folder though).

**Reply by krivulcik** [reply]
I sent the raw database files to support email. On each node, I stopped the RavenDB service, zipped the database directory, and then started the RavenDB service. I removed the Indexes directory from the zip afterwards.

**Reply by lastav5** [reply]
Following your email to support, the issue has been reproduced:

Node A suffered hardware failure and restarted. Before it started receiving all documents from the other nodes, it executed nodes B and C's awaiting cluster transactions. Node A now had documents with a change vector of type 'Raft-x'. Once replication from B and C started happening, all incoming documents that had a change vector of type 'Raft-x' were lower than the existing ones on node A and were rejected.

This later happened on node B as well, resulting in both nodes A and B missing documents.

A workaround for this is enabling the atomic guards. When they are enabled, the TRXN field joins the document's change vector (so now each cluster transaction doc's cv is of type 'RAFT-x, TRXN-y'). However, the database's change vector doesn't contain the TRXN field, so replicating a TRXN item will always result in a conflict (between the database cv and the item cv), which in turn results in accepting the item.

Issue has been opened: https://issues.hibernatingrhinos.com/issue/RavenDB-18937
